### PR TITLE
Introduce Shadow-Pilot

### LINKED ISSUE #861: Autopilot shadow mode
We need to be able to run our autopilot in shadow mode so that we can test new solvers without submitting anything onchain or disrupting the existing auction. This shadow mode should allow the autopilot to fetch auctions from e.g. staging or production without affecting them in any way. It also needs to fetch the solutions generated by other solvers on staging or production and run those solutions though the ranking process, along with the solutions of the new solver being tested.

1. Add a `--shadow <URL>` flag to autopilot. This flag is only valid if the submission strategy is `DryRun`.
2. Do not cut batches, instead get batches from `<URL>/api/v1/auction`.
3. When solving, wait for `<URL>/api/v1/solver_competition/{auction_id}` to return a valid response. Merge the solutions from that endpoint with the solutions of the configured solvers.
4. Do the remaining settlement rating, storing the results, etc. as usual. Note that since the strategy must be `DryRun` nothing is published onchain.

Blocked until https://github.com/cowprotocol/services/issues/862 is done.

### PR DESCRIPTION
**Pull request by nlordell**
Fixes #861

This PR adds an autopilot shadow mode for co-location.

It works by polling a public CoW Protocol API endpoint for auctions, and does the competition for all configured solvers.

Logs competition results and metrics including:
- The for the winning driver and solution and its approximate reward (assuming that the proposed solution score == observed value, which isn't true). Note we can probably use Tenderly to simulate an observed score on the current block, and then compute the actual reward, but that is additional work and wouldn't fit into this (already large) PR.
- The calldata and score for all participants
- Metric for the last observed auction (so we can setup alerts if the shadow competition goes down for example)
- Metrics for the total rewards per solver (again, with a "shadow dashboard" in mind)
- Metrics for participation (also with a "shadow dashboard" in mind)

### Test Plan

Run locally, see competition happening!

```
% cargo run -p autopilot -- --shadow https://api.cow.fi/goerli --drivers http://localhost:11088/mybologna
...
2023-09-09T12:16:15.475Z  INFO auction{id=8842306}: autopilot::shadow: winner driver=http://localhost:11088/mybologna score=27744239924793385218 reward=27744239924793385218
2023-09-09T12:16:15.476Z DEBUG auction{id=8842306}: autopilot::shadow: participant driver=http://localhost:11088/mybologna score=27744239924793385218 account=0xdbc682d0ec19bd276fd6c2af4e0189b584bb31e1 calldata=0x13d79a0b000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000360000000000000000000000000000000000000000000000000000000000000000200000000000000000000000091056d4a53e1faa1a84306d4deaec71085394bc8000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d6000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000019211a3b3ba2916e00000000000000000000000000000000000000000000010f0cf064dd592000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000dbc682d0ec19bd276fd6c2af4e0189b584bb31e100000000000000000000000000000000000000000000000019616ef94442e0f700000000000000000000000000000000000000000000010f0cf064dd592000000000000000000000000000000000000000000000000000000000000064fc6907e5fe9a8503b38e3b55abfdeef9454b86e03b1e54ad0c2d819ce056ed8ce6a6020000000000000000000000000000000000000000000000000000000022fc8031000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000010f0cf064dd5920000000000000000000000000000000000000000000000000000000000000000001600000000000000000000000000000000000000000000000000000000000000041ea810d80ada9dcc0d69d823e702e5f94a254ebc7dc0fb911bc14314834844fec14913f8c3197dfbd397af54dfb9c93938335c11434deb309137811978e2094a21b000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000002600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000200000000000000000000000007a250d5630b4cf539739df2c5dacb4c659f2488d0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000001048803dbee00000000000000000000000000000000000000000000010f0cf064dd592007d60000000000000000000000000000000000000000000000001ba469a78e66066000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000009008d19f58aabd9ed0d60971565aa8510560ab41ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d600000000000000000000000091056d4a53e1faa1a84306d4deaec71085394bc8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000086ec42
...
```

### CODE REVIEW

File: crates/autopilot/src/run_loop.rs [hunk 1]
```diff
@@ -269,75 +269,8 @@ impl RunLoop {
             return Default::default();
         }
 
-        let request = &solve::Request {
```
Comment (nlordell) [inline crates/autopilot/src/run_loop.rs:272]: Code just moved down with minimal changes.

File: crates/autopilot/src/shadow.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+                        driver =% driver.url,
+                        score =% solution.score,
+                        account =? solution.account,
+                        calldata =% hex(&solution.calldata.internalized),
+                        "participant"
+                    );
+                    if solution.calldata.internalized != solution.calldata.uninternalized {
+                        tracing::debug!(
+                            driver =% driver.url,
+                            uninternalized =% hex(&solution.calldata.uninternalized),
```
Comment (MartinquaXD) [inline crates/autopilot/src/shadow.rs:144]: Would be nice if this could be folded into the other log.

File: crates/autopilot/src/xapi/mod.rs [hunk 3]
```diff
@@ -0,0 +1,5 @@
+//! A top-level module foe client implementations to e**X**ternal **API**s.
```
Comment (MartinquaXD) [inline crates/autopilot/src/xapi/mod.rs:1]:
```suggestion
//! A top-level module for client implementations to e**X**ternal **API**s.
```
Also the name `xapi` seems unfortunate IMO. I didn't get what it's supposed to mean until I read this comment.
Additionally strictly speaking you wouldn't have to use it with an external API. To me this seems like a client for the public API. It's up to the user of the tool if the provided API is external or actually the one in the same cluster as the `autopilot`.

File: crates/number/src/nonzero.rs [hunk 4]
```diff
@@ -1,20 +1,36 @@
...
 #[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
 pub struct U256(ZeroU256);
+
+impl U256 {
+    pub fn new(value: ZeroU256) -> Option<Self> {
+        if !value.is_zero() {
+            Some(Self(value))
+        } else {
+            None
+        }
```
Comment (MartinquaXD) [inline crates/number/src/nonzero.rs:17]: I just have to point out this hidden gem every time I see it. :sweat_smile:
```suggestion
        !value.is_zero().then_some(Self(value))
```

File: crates/autopilot/src/shadow.rs [hunk 5]
```diff
@@ -0,0 +1,270 @@
+use {
```
Comment (MartinquaXD) [inline crates/autopilot/src/shadow.rs:1]: A doc comment for this module would be nice (basically same info that's already part of the CLI `--help` text).

**Review by MartinquaXD — APPROVED** [review]

**Comment by sunce86** [comment]
IIUC the code, once we start shadow autopilot for any new solver, we will be sending duplicate `/solve` requests to all drivers?

File: crates/number/src/nonzero.rs (hunk 4 above)
Comment (nlordell) [reply crates/number/src/nonzero.rs:17]: I always forget...

**Comment by nlordell** [comment]
The idea is to have a single `shadow` autopilot communicating with all drivers.

File: crates/autopilot/src/xapi/mod.rs (hunk 3 above)
Comment (nlordell) [reply crates/autopilot/src/xapi/mod.rs:1]: I hate the name `xapi`... just didn't know what else to name this module... Let me try something else.

File: crates/autopilot/src/shadow.rs (hunk 5 above)
Comment (nlordell) [reply crates/autopilot/src/shadow.rs:10]: c567f420c2b438976e12ace04484258a670c7577

File: crates/autopilot/src/shadow.rs (hunk 2 above)
Comment (nlordell) [reply crates/autopilot/src/shadow.rs]: c567f420c2b438976e12ace04484258a670c7577

Comment (nlordell) [reply crates/autopilot/src/shadow.rs]: > Also the log says `internalized interactions` when it prints the `uninternalized` data.

This was intentional - the log was indicating that the solver chose to internalize interactions, and then prints the full uninternalized calldata in the message. Makes sense to me at least 😅.

Comment (MartinquaXD) [reply crates/autopilot/src/shadow.rs]: Ahh, I see. Seemed a bit unintuitive to me but I can see where you are coming from now.

File: crates/autopilot/src/xapi/mod.rs (hunk 3 above)
Comment (MartinquaXD) [reply crates/autopilot/src/xapi/mod.rs]: I also don't have a great suggestion from the top of my head but I think the new name is an improvement. 👍

File: crates/autopilot/src/arguments.rs [hunk 6]
```diff
@@ -159,6 +159,15 @@ pub struct Arguments {
     /// is still considered for payout.
     #[clap(long, env, default_value = "5")]
     pub additional_deadline_for_rewards: usize,
+
+    /// Run the autopilot in a shadow mode by specifying an upstream CoW
+    /// protocol deployment to pull auctions from. This will cause the autopilot
+    /// to start a run loop where it performs solver competition on driver,
+    /// and report and log the winner **without** actually executing any
```
Comment (fleupold) [inline crates/autopilot/src/arguments.rs:166]: The autopilot doesn't execute settlements in its normal mode neither, does it?

File: crates/autopilot/src/run.rs [hunk 7]
```diff
@@ -590,3 +597,48 @@ pub async fn run(args: Arguments) {
...
+
+async fn shadow_mode(args: Arguments) -> ! {
+    let http_factory = HttpClientFactory::new(&args.http_client);
+
+    let orderbook = protocol::Orderbook::new(
+        http_factory.create(),
+        args.shadow.expect("missing shadow mode configuration"),
+    );
+
+    let drivers = args.drivers.into_iter().map(Driver::new).collect();
```
Comment (fleupold) [inline crates/autopilot/src/run.rs:609]: Should we somehow assert that colocation is enabled and drivers is non empty as we do fort the regular runloop?

File: crates/autopilot/src/run.rs [hunk 8]
```diff
@@ -87,6 +89,11 @@ pub async fn start(args: impl Iterator<Item = String>) {
 
 /// Assumes tracing and metrics registry have already been set up.
 pub async fn run(args: Arguments) {
+    if args.shadow.is_some() {
```
Comment (fleupold) [inline crates/autopilot/src/run.rs:92]: nanonit: maybe move up into the main function to keep the original run cleaner?

**Review by fleupold — APPROVED** [review]
Nice!

File: crates/autopilot/src/run.rs (hunk 7 above)
Comment (nlordell) [reply crates/autopilot/src/run.rs:619]: Added assert. Explicitly documented that `--enable-colocation` does not need to be set and is assumed in `--shadow` mode.

File: crates/autopilot/src/arguments.rs (hunk 6 above)
Comment (nlordell) [reply crates/autopilot/src/arguments.rs]: No. I clarified this in the documentation.

**Pull request merged**